tests/backends/test_base.py

```diff
@@ -0,0 +1,87 @@
+import importlib
+
+import pytest
+
+from cohortextractor.backends.base import BaseBackend, Column, MappedTable
+from cohortextractor.concepts import types
+from cohortextractor.concepts.table_contract import BackendContractError
+from cohortextractor.concepts.table_contract import Column as ColumnContract
+from cohortextractor.concepts.table_contract import TableContract
+from cohortextractor.query_engines.base_sql import BaseSQLQueryEngine
+
+
+class PatientsContract(TableContract):
+    """
+    A test contract that backends can be validates against
+    """
+
+    patient_id = ColumnContract(type=types.PseudoPatientId(), help="")
+    date_of_birth = ColumnContract(type=types.Date(), help="")
+    sex = ColumnContract(type=types.Choice("F", "M"), help="")
+
+
+def test_bad_backend_with_validate_contract():
+
+    # Create a Backend which will fail to implement a contract
+    class BadBackend(BaseBackend):
+        backend_id = "bad_test_backend"
+        query_engine_class = BaseSQLQueryEngine
+        patient_join_column = "patient_id"
+
+        patients = MappedTable(
+            implements=PatientsContract,
+            source="Patient",
+            columns=dict(
+                date_of_birth=Column("date", source="DateOfBirth"),
+            ),
+        )
+
+    bad_backend = BadBackend(database_url="test")
+
+    with pytest.raises(BackendContractError, match="Missing columns: sex"):
+        bad_backend.validate_all_contracts()
+
+
+def test_good_backend_with_validate_contract():
+    # Create a Backend which will implement a contract
+    class GoodBackend(BaseBackend):
+        backend_id = "good_test_backend"
+        query_engine_class = BaseSQLQueryEngine
+        patient_join_column = "patient_id"
+
+        patients = MappedTable(
+            implements=PatientsContract,
+            source="Patient",
+            columns=dict(
+                date_of_birth=Column("date", source="DateOfBirth"),
+                sex=Column("varchar", source="Sex"),
+            ),
+        )
+
+    good_backend = GoodBackend(database_url="test")
+    good_backend.validate_all_contracts()
+
+
+def test_validate_all_backends():
+    """
+    Loops through all the backends, excluding test ones,
+    and validates they meet any contract that they claim to
+    """
+    real_backends = [cls.__name__ for cls in BaseBackend.__subclasses__()]
+    dummy_backends = ["BadBackend", "GoodBackend"]
```

Comment:
Thinking about this again, I have doubts about my suggestion that it's better to do this by exclusion. If someone adds a new test backend there's no reason for them to think of excluding it here. How about we get all the subclasses and then only check the ones that are defined in `cohortextractor.backends`? We might add an assertion that there were at least three of them, to make sure we don't run into a problem with package restructuring.

Reply:
Great - I have done this by removing reference to the dummy backends. The `getattr()` takes care of finding the real backends.